### ISSUE TITLE
reflection: message descriptor doesn't not get generated if the message is not used in code

### ISSUE THREAD

**Issue by kelvinh** [google/protobuf#2104]
Hi all, I am experiencing a weird issue, I use reflection to create message instance in C++, as the following code shows:

```
std::string msg_name = "foo";
const Descriptor *desc = DescriptorPool::generated_pool()->FindMessageTypeByName(msg_name);
assert(desc);
Message *msg = MessageFactory::generated_factory()->GetPrototype(desc)->New();
assert(msg);
```

Assume the proto file is "test.proto", and just includes two messages, foo and bar:

```
message foo { ... }
message bar { ... }
```

Then, the C++ code fails at `assert(desc)` because desc is NULL. However, if I add the following code at **any place** in my code, the desc is not NULL:

```
foo f;
(void) f;
```

Also, declaration an instance of bar will also get the code work:

```
bar b;
(void) b;
```

I get suprised by this, I don't know if there is some black magic here... Is there any way to avoid this?

I am using protobuf 2.6.1 with statically linking, g++ 5.4.0 on Ubuntu 16.04.

**Comment by kelvinh** [comment]
Why I call it **black magic**, because the instance declaration never get called:

```
int main(int argc, char **argv) {
    if (argc > 10000) {
        bar b;
        (void) b;
    }
    ...
}
```

I also noticed the following generated code in "test.pb.cc":

```
void protobuf_AddDesc_5ftest_2eproto() {
  static bool already_here = false;
  if (already_here) return;
  already_here = true;
  GOOGLE_PROTOBUF_VERIFY_VERSION;

   // reflection initialization code here, omitted
}

// Force AddDescriptors() to be called at static initialization time.
struct StaticDescriptorInitializer_5ftest_2eproto {
  StaticDescriptorInitializer_5ftest_2eproto() {
    printf("static ctor called\n");                   // <-- I add this line, not generated by protoc
    protobuf_AddDesc_5ftest_2eproto();
  }
} static_descriptor_initializer_5ftest_2eproto_;
```

the static struct global variable's constructor should get called before `main(...)` to add descriptors to the generated descriptor pool, but, the line I added never get printed if I did not add the code `bar b; (void) b;`...

**Comment by kelvinh** [comment]
OK, finally I figured out that there is no black magic and it's just a linker problem, I compiled the proto sources into a static library, seems the linker automatically optimized out those objs not used when linking against the static library. I follow [this](http://stackoverflow.com/questions/4767925/how-to-force-gcc-to-link-unreferenced-static-c-objects-from-a-library) thread and it fixes the problem.

Sorry to disturb you all for the inconvenience. :-)
